### ISSUE TITLE
question about the explaination of the code

### ISSUE THREAD

**Issue by supernihui** [experiencor/basic-yolo-keras#24]
Hi,author, I have two questions for your code.
1.I have used your code on the Raccoon dataset which you provided. But as you have not mentioned much details, I have used the pretrained weights of tiny yolo and found no box detected no matter what raccoon image I found on the internet or I select from the database you provided.

2.In your model.py code,  pred_box_xy = tf.sigmoid(y_pred[..., :2]) + cell_grid. In my opinion, this code just calculate the paper bx=sigma(tx)+cx, but the cell_grid is fixed matrix not some variables, and will not change with different location of the cell.I can't understand.
Thanks

**Comment by experiencor** [comment]
1. I had the same experience as yours. The tiny version of yolo does not work out of the box with the racoon dataset and you have to play the parameter coord_scale in the config.json file. You can try out this strategy:

+ python train.py with warmup_batches = 100, train_times = 10, and coord_scale = 1
+ python train.py again with warmup_batches = 0, train_times = 10, and coord_scale = 3

The reason of doing like this is both theoretical and empirically. But you will get the hang of it when trying out different parameters.

P/S: the full version will work out of the box.

2. cell_grid is the 2D version of range(n). It contains the offsets of the cells.Cells further from top left corner have high offsets. But the offsets are fixed. Why do you understand that they should vary?

**Comment by supernihui** [comment]
Thanks and I have understood your answer for question 1. 
I  read the formula in the paper, and compare with that formula. I thought cell_grid in your code should just be the Cx and then I thought it should be a variable. And I didn't see the original code because it was written on C.  I don't know how the author realize the cx in his code and the cx is not covered how to implement in his paper.

**Comment by experiencor** [comment]
You are right. cell_grid is exactly (cx,cy). It's fixed for any position inside a cell. But it's different for different cells.

**Comment by Pezaun** [comment]
Hi everybody!

Experiencor, can you help me to understand your code? I have 2 questions:

1 - About Warm-up code present in your loss function. Why we have to warn-up in each iteration?
2 - Why you divide nb_coord_box and nb_conf_box by 2? This intuitions are not present in the papers. Did you found it in darknet code?

Congratulations for your work! Thanks!

**Comment by supernihui** [comment]
Thanks for your answer!!!!

**Comment by experiencor** [comment]
1. According to the 2 YOLO papers, the boxes which are not responsible for predictions don't have their coordinates penalized. At the start, as the weights are random, the predictions of these boxes could be anything. It's not good as we want the boxes to make predictions near to the sizes of the anchors, and so the objects. To address this issue, during initial batches, we actually penalize these boxes to force them to make predictions near to the sizes of the anchors. 

2. I copy it from the pytorch implementation (https://github.com/marvis/pytorch-yolo2), which copies from the original darknet code. The mean squared error usually comes with the coefficient 1/2 to simplify the math of derivative (think $d(x^2/2)/dx=x$). In my opinion, it's just a convenient convention and not essential.

**Comment by Pezaun** [comment]
Thank you for the answer! I'm going nuts to implement my own code. It can overfit with some images, but it does not generalize, even with data augmentation. I can only train with SGD, when I train it with ADAM my loss become nan in a few steps. :/

Can you take a look at my loss function? (https://github.com/Pezaun/GOLO/blob/master/golo_loss.py)

Thanks

**Comment by experiencor** [comment]
@Pezaun I also had the NAN problem with the previous version of my implementation. I did not know how to solve it. But this problem seems to disappear after I add the warming training code. I think that it has something to do stability during initial batches. 

BTW, which loss (xy, wh, confidence, ...) do your NANs come from?

**Comment by Pezaun** [comment]
@experiencor , thanks for your answer! The NANs come from wh. I observe that in some moment tf.exp(pred_wh) became inf. I will try to implement the warm-up as you make. My objective with this implementation is the multicale training, but it have to learn for static scale first.

Thank you again!

**Comment by experiencor** [comment]
You may refer to https://www.tensorflow.org/api_docs/python/tf/space_to_depth. This operation can be implemented by a chain of reshape and swap axis operations. tf.space_to_depth is used to keep things simple.

**Comment by thorstenwagner** [comment]
@experiencor 
"According to the 2 YOLO papers, the boxes which are not responsible for predictions don't have their coordinates penalized. At the start, the predictions of these boxes could be anything as the weights are random. This is not a good start. We do want these boxes to make predictions near to the sizes of the anchors so that when they are to be responsible for some object, they only have to slightly change their predictions to fit the size of that object. To address this issue, during initial batches, we do penalize these boxes to force them to make predictions near to the sizes of the anchors."

Could you tell use section of the paper where it is written? I cant find the description of the warm up phase.
